src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,464 @@
+                {
+                    continue;
+                }
+
+                if (!item.Properties.TryGetProperty(CompletionSource.RoslynItem, out RoslynCompletionItem roslynItem))
+                {
+                    roslynItem = RoslynCompletionItem.Create(
+                        displayText: item.DisplayText, 
+                        filterText: item.FilterText, 
+                        sortText: item.SortText,
+                        displayTextSuffix: item.Suffix);
+                }
+
+                if (Session.MatchesFilterText(_completionHelper, roslynItem, filterText, initialRoslynTriggerKind, filterReason, _recentItems))
+                {
+                    initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: true)));
+                }
+                else
+                {
+                    // The item didn't match the filter text.  We'll still keep it in the list
+                    // if one of two things is true:
+                    //
+                    //  1. The user has only typed a single character.  In this case they might
+                    //     have just typed the character to get completion.  Filtering out items
+                    //     here is not desirable.
+                    //
+                    //  2. They brough up completion with ctrl-j or through deletion.  In these
+                    //     cases we just always keep all the items in the list.
+                    if (roslynTrigger.Kind == CompletionTriggerKind.Deletion ||
+                        roslynTrigger.Kind == CompletionTriggerKind.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false)));
+                    }
+                }
+            }
+
+            // If the session was created/maintained out of Roslyn, e.g. in debugger; no properties are set and we should use data.Snapshot.
+            // However, we prefer using the original snapshot in some projection scenarios.
+            if (!session.Properties.TryGetProperty(CompletionSource.TriggerSnapshot, out ITextSnapshot snapshotForDocument))
+            {
+                snapshotForDocument = data.Snapshot;
+            }
+
+            var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
+            var completionService = document?.GetLanguageService<CompletionService>();
+            var completionRules = completionService?.GetRules() ?? CompletionRules.Default;
+
+            if (data.Trigger.Reason == CompletionTriggerReason.Backspace &&
+                completionRules.DismissIfLastCharacterDeleted &&
+                session.ApplicableToSpan.GetText(data.Snapshot).Length == 0)
+            {
+                // Dismiss the session
+                return null;
```

Comment:
Earlier in this method we did session.Dismiss() before returning null, what's the difference?